### PR TITLE
perf(CD): Special cased interpolation in AST, Parser, and CD

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: modules/change_detection/src/parser/parser.js [hunk 1]
```diff
@@ -52,6 +56,29 @@ export class Parser {
     var tokens = this._lexer.tokenize(input);
     return new _ParseAST(input, location, tokens, this._reflector, false).parseTemplateBindings();
   }
+
+  parseInterpolation(input:string, location:any):ASTWithSource {
+    var parts = StringWrapper.split(input, INTERPOLATION_REGEXP);
+    if (parts.length <= 1) {
+      return null;
+    }
+    var strings = [];
```
Comment (vicb) [inline modules/change_detection/src/parser/parser.js]: Creating fixed length lists might be more efficient (we already know the length)

File: modules/change_detection/src/record_range.js [hunk 2]
```diff
@@ -546,6 +557,38 @@ function _mapFn(keys:List, length:int) {
...
+  var c0 = length > ++i ? strings[i] : null;
+  var c1 = length > ++i ? strings[i] : null;
+  var c2 = length > ++i ? strings[i] : null;
+  var c3 = length > ++i ? strings[i] : null;
+  var c4 = length > ++i ? strings[i] : null;
+  var c5 = length > ++i ? strings[i] : null;
+  var c6 = length > ++i ? strings[i] : null;
+  var c7 = length > ++i ? strings[i] : null;
+  var c8 = length > ++i ? strings[i] : null;
+  var c9 = length > ++i ? strings[i] : null;
```
Comment (vicb) [inline modules/change_detection/src/record_range.js]: I'm curious why a variable is used over constants, is this more efficient ?

**Comment by vicb** [comment]
A couple of minor comments otherwise LGTM

File: modules/change_detection/src/parser/parser.js (hunk 1 above)
Comment (mhevery) [inline modules/change_detection/src/parser/parser.js]: Turns out for small arrays, there is no, or slight favor for not declaring, for large arrays there is a big difference.

File: modules/change_detection/src/record_range.js (hunk 2 above)
Comment (mhevery) [inline modules/change_detection/src/record_range.js]: It is simpler to read/code, has no perf difference since the V8 will keep it in a local register.

**Comment by mhevery** [comment]
Closed by: 3b34ef43b185609b307199e53370612b0e12f7b4

**Pull request closed without merging**